mypyc/codegen/emitclass.py

```diff
@@ -698,13 +720,22 @@ def generate_traverse_for_class(cl: ClassIR,
             emitter.emit_gc_visit(f'self->{emitter.attr(attr)}', rtype)
     if cl.has_dict:
         struct_name = cl.struct_name(emitter.names)
-        # __dict__ lives right after the struct and __weakref__ lives right after that
-        emitter.emit_gc_visit('*((PyObject **)((char *)self + sizeof({})))'.format(
-            struct_name), object_rprimitive)
-        emitter.emit_gc_visit(
-            '*((PyObject **)((char *)self + sizeof(PyObject *) + sizeof({})))'.format(
-                struct_name),
-            object_rprimitive)
+        # TODO: Use PY_VERSION_HEX and include patchlevel.h
+        if emitter.capi_version < (3, 11):
+            # __dict__ lives right after the struct and __weakref__ lives right after that
+            emitter.emit_gc_visit('*((PyObject **)((char *)self + sizeof({})))'.format(
+                struct_name), object_rprimitive)
+            emitter.emit_gc_visit(
+                '*((PyObject **)((char *)self + sizeof(PyObject *) + sizeof({})))'.format(
+                    struct_name),
+                object_rprimitive)
+        else:
+            # In >=3.11, __dict__ is managed by CPython and __weakref__ lives right after
+            # the struct
+            emitter.emit_gc_visit(
```

Comment:
Will this traverse `__dict__`? CPython relies on `subtype_traverse` in `typeobject.c` to traverse the dict, not sure how mypyc creates heap types.

Reply:
Wait, does `subtype_traverse` not apply to this type?

Reply:
I just checked the flags:
```
flags = ['Py_TPFLAGS_DEFAULT', 'Py_TPFLAGS_HEAPTYPE', 'Py_TPFLAGS_BASETYPE']
(and `Py_TPFLAGS_MANAGED_DICT`)
```
So this means the traverse function will automatically detect the dict and visit it during GC right?

Reply:
`subtype_traverse` is only added for classes which are created by `type.__call__` and the  type metaclass's `tp_new`  but mypyc does it's custom class creation thing and does not seem to call `type()` so I am not sure.

Reply:
Ohh thanks you're right. Hmm in that case it's probably not being GC-ed then.

Reply:
After consulting Mark, it appears that 3rd party code setting `Py_TPFLAGS_MANAGED_DICT` doesn't really have a safe way to GC its own dict.

I can think of two possible solutions:
1. Use the private `_PyObject_VisitInstanceAttributes` by including `pycore_object.h`. OR
2. Manage the dict ourselves.